target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c

```diff
@@ -0,0 +1,484 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/module.h>
+#include <linux/of_platform.h>
+#include "i2c-rtl9300.h"
+
+#define REG(x)		(i2c->base + x + (i2c->scl_num ? 0x1c : 0))
+#define REG_MASK(clear, set, reg)	\
+			writel((readl(REG(reg)) & ~(clear)) | (set), REG(reg))
+
+struct i2c_drv_data {
+	int scl0_pin;
+	int scl1_pin;
+	int sda0_pin;
+	struct i2c_algorithm *algo;
+	int (*read)(struct rtl9300_i2c *i2c, u8 *buf, int len);
+	int (*write)(struct rtl9300_i2c *i2c, u8 *buf, int len);
+	void (*reg_addr_set)(struct rtl9300_i2c *i2c, u32 reg, u16 len);
+	int (*config_xfer)(struct rtl9300_i2c *i2c, u16 addr, u16 len);
+	int (*execute_xfer)(struct rtl9300_i2c *i2c, char read_write, int size,
+			    union i2c_smbus_data * data, int len);
+	void (*writel)(struct rtl9300_i2c *i2c, u32 data);
+	void (*config_io)(struct rtl9300_i2c *i2c, int scl_num, int sda_num);
+};
+			
+DEFINE_MUTEX(i2c_lock);
+
+static void rtl9300_i2c_reg_addr_set(struct rtl9300_i2c *i2c, u32 reg, u16 len)
+{
+	// Set register address width
+	REG_MASK(0x3 << RTL9300_I2C_CTRL2_MADDR_WIDTH, len << RTL9300_I2C_CTRL2_MADDR_WIDTH,
+		 RTL9300_I2C_CTRL2);
```

Comment:
I really prefer a few #define-s to document what the register fields are, and then using FIELD_PREP, FIELD_GET,... from bitfield.h.

Reply:
RTL9300_I2C_CTRL2_MADDR_WIDTH is anyway defined. FIELD_PREP and FIELD_GET will just increase the line-count considerably here. The line says quite a lot: RTL9300_I2C_CTRL2_MADDR_WIDTH is a 2-bit field starting at RTL9300_I2C_CTRL2_MADDR_WIDTH, which is really not more than you can anyway get. I'll leave this to the taste of a subsystem maintainer should this be submitted.

Reply:
Line count by itself is a bad metric. Every next person will have to figure out and understand what this macro means. Using FIELD_PREP is not so bad really:

    u32 v;
    v = readl(RTL9300_IC2_CTRL2) & ~RTL9300_IC2_CTRL_MADDR;
    writel(v | FIELD_PREP(RTL9300_I2C_CTRL_MADDR, len), RTL9300_I2C_CTRL2);

Or if you really want to use a macro:

    #define REG_FIELD_WRITE(offset, mask, val) \
        writel((readl(offset) & ~mask) | FIELD_PREP(mask, val), offset)
    REG_FIELD_WRITE(RTL9300_I2C_CTRL2, RTL9300_I2C_CTRL_MADDR, len);

IMHO `REG_MASK` also just isn't a good name for a macro that clears and sets two arbitrary sets of bits.